pincer/core/gateway.py

```diff
@@ -46,275 +66,384 @@ class Dispatcher:
+        # be tracked.
+        if payload.seq is not None:
+            self.__sequence_number = payload.seq
+            _log.debug("%s Set sequence number to %s", self.shard_key, payload.seq)
+
+        handler = self.__dispatch_handlers.get(payload.op)
+
+        if handler is None:
+            raise UnhandledException(f"Opcode {payload.op} does not have a handler")
 
-        execute_handler(handler)
-        execute_handler(all_handler)
+        ensure_future(handler(payload))
 
-    async def __dispatcher(self, loop: AbstractEventLoop):
+    async def handle_heartbeat_req(self, payload: GatewayDispatch):
         """
-        The main event loop.
-        This handles all interactions with the Websocket API.
+        Opcode 1 - Instantly send a heartbeat.
+        """
+        self.send_next_heartbeat()
 
-        Parameters
-        ----------
-        loop : :class:`~asyncio.AbstractEventLoop`
-            The loop in which the dispatcher is running.
+    async def handle_reconnect(self, payload: GatewayDispatch):
+        """
+        Opcode 7 - Reconnect and resume immediately.
         """
         _log.debug(
-            "Establishing websocket connection with `%s`", GatewayConfig.uri()
+            "%s Requested to reconnect to Discord. Closing session and attempting to"
+            "resume...",
+            self.shard_key
         )
 
-        async with connect(GatewayConfig.uri()) as socket:
-            socket: WebSocketClientProtocol = socket
-            self.__socket = socket
+        await self.__socket.close(code=1000)
+        self.__should_reconnect = True
+        await self.start_loop()
+
+    async def handle_invalid_session(self, payload: GatewayDispatch):
+        """
+        Opcode 9 - Invalid connection
+        Attempt to relog. This is probably because the session was already invalidated
+        when we tried to reconnect.
+        """
+        _log.debug("%s Invalid session, attempting to relog...", self.shard_key)
+        self.__should_reconnect = False
+        await self.start_loop()
+
+    async def identify_and_handle_hello(self, payload: GatewayDispatch):
+        """
+        Opcode 10 - Hello there general kenobi
+        Runs when we connect to the gateway for the first time and every time after.
+        If the client thinks it should reconnect, the opcode 6 resume payload is sent
+        instead of the opcode 2 hello payload. A new session is only started after a
+        reconnect if pcode 9 is received.
+
+        Successful reconnects are handled in the `resumed` middleware.
+        """
+        if self.__should_reconnect:
+            _log.debug("%s Resuming connection with Discord", self.shard_key)
+
+            await self.send(str(GatewayDispatch(
+                6,
+                {
+                    "token": self.token,
+                    "session_id": self.__session_id,
+                    "seq": self.__sequence_number
+                }
+            )))
```

Comment:
Stay consistent w/ the one below
```suggestion
            await self.send(str(GatewayDispatch(
                6, {
                    "token": self.token,
                    "session_id": self.__session_id,
                    "seq": self.__sequence_number
                }
            )))
```